batch/batch/front_end/validate.py

```diff
@@ -8,6 +8,7 @@
 #   'always_run': bool,
 #   'attributes': {str: str},
 #   'command': [str],
+#   'entrypoint': str or list[str]
```

Comment:
My proposal is backend-only and shouldn't involve changing the Python client library or REST API.

Wait, is your problem the docker image has an `ENDPOINT` already defined, or that the docker image doesn't have `/bin/bash` installed?

I see `/bin/bash` is still hardcoded.  I thought this had been fixed, this is still a pending Asana item.  The solution here is that the `Batch` constructor should take a default shell option that can be overridden per-job.  This should be used by the backend, defaulting to `/bin/bash` if not specified.  This design overlaps somewhat with what you've proposed.

Reply:
ENTRYPOINT defined. Goal is to unset. Since you all don't have any interest in entrypoint being threaded through, I can just set Entrypoint="" in the worker

Reply:
The 2nd thing my PR allowed was handling non /bin/bash environments. I thought you said you don't want that, so removed ability to set it. That was effectively handled with this PR

Reply:
I'm not asking for a use case, I'm asking for an interface.  Let me be more specific:

 - The `Batch` constructor takes a `shell: str = None` keyword argument,
 - `Batch.new_job` takes a `shell: str = None` keyword argument that overrides the Batch argument,
 - the job spec in the service backend has a new optional `shell: str` field,
 - the backends use shell, if given, instead of `/bin/bash` on a per-Job basis.

Reply:
>  I'm asking for an interface.

The interface: 

* Not needed: The Batch constructor changes
* Not needed: batch.new_job changes
* job.entrypoint(arg: Union(str, List[str]), job spec has a new optional entrypoint. 
* The backend uses entrypoint (I will need a small change to this PR to make it override /bin/bash in service backend)

Reply:
I'm asking for the interface I wrote down.  I thought I said clearly: our interface is not meant to be a wrapper for Docker and endpoint is the wrong concept here.

Reply:
Why does the command I wrote need to have anything to do with Docker? Call the method whatever you want. The current backend is Docker, but that's an implementation detail.  I wrote what you wrote except pointed out that in this PR the first two things you wrote are not necessary.

Reply:
Also I don't know what "I'm asking for the interface I wrote down means" in the context of your earlier suggestion that there may be overlap with this implementation except a mapping.

Reply:
Let be be as clear as I can: change this PR to remove entrypoint from the interface and implement this interface instead:

 > The Batch constructor takes a shell: str = None keyword argument,
 > Batch.new_job takes a shell: str = None keyword argument that overrides the Batch argument,
 > the job spec in the service backend has a new optional shell: str field,
 > the backends use shell, if given, instead of /bin/bash on a per-Job basis.

Also unset the entrypoint with `--entrypoint ""` in the local backend to fix your original issue.

Reply:
As an aside, your proposed interface is called "shell". I don't like that. Everyone who uses Batch will think about Docker. Docker has a primitive for overriding a command. Why not use something that has the same name if it does almost the same thing (yours is a subset of the Docker ENTRYPOINT interface). 

People have enough difficulty understanding how to containerize things. We're now going to add another term for the same thing. So TJ will try to containerize something in Docker. He sees that there is something in Docker called Entrypoint. He will realize that Job command has a bunch of magic behind the scenes that makes it impossible to supply the command that he wants to his image. Now we wants to override ENTRYPOINT. He'll search for that. Nothing. He has no idea what various "shell"s are. This is a computer science artifact. 

Better to name the thing what is expected. This has nothing to do with Batch having a docker backend, or a cgroup backend. You'll need some way to go from a docker image to executes code. My 2c.

Reply:
> Let be be as clear as I can: change this PR to remove entrypoint from the interface and implement this interface instead:
> 
> 
> 
>  > The Batch constructor takes a shell: str = None keyword argument,
> 
>  > Batch.new_job takes a shell: str = None keyword argument that overrides the Batch argument,
> 
>  > the job spec in the service backend has a new optional shell: str field,
> 
>  > the backends use shell, if given, instead of /bin/bash on a per-Job basis.
> 
> 
> 
> Also unset the entrypoint with `--entrypoint ""` in the local backend to fix your original issue.
> 
> 

You said this was per-job. Why make it per batch? You specify images per job, commands, they may have different entry point needs.

Reply:
I'm telling you, we already decided not to do that.  `shell` isn't entrypoint by a different name.  `shell` is literally the shell used to interpret the shell script constructed from the job command.  What you are proposing will not work.  If you want to run plink, setting the shell or entrypoint to plink will not work: it does not take `-c <script>` as valid arguments.  We use this for several features: `verbose` adds `set +x` to the script, running `command` multiple times creates a shell script with `&&`, etc.  What you are proposing is incorrect in the context of the current design which I am trying to describe.

I understand you dislike the existing design.  The options I see are:
 - you just unset `entrypoint` for this PR and we sort out the interface question separately,
 - you implement the interface we've created to work with the existing design, or
 - you close this and make a larger proposal to replace our existing interface with one modelled on the docker interface and discuss.

Reply:
> You said this was per-job. Why make it per batch? You specify images per job, commands, they may have different entry point needs.

No, there is a default image set on the batch.  The jobs can override that.  I'm asking for exactly the existing interface for the shell.  (It is not an endpoint.  See my other comment.)  In fact, that's true for many of the Job attributes:

```
class Batch:
    ...
    def __init__(self,
                 ...
                 default_image: Optional[str] = None,
                 default_memory: Optional[str] = None,
                 default_cpu: Optional[str] = None,
                 default_storage: Optional[str] = None,
                 default_timeout: Optional[Union[float, int]] = None):
```

Reply:
good point, thanks

Reply:
Let me know if current state of PR is what you're looking for. There's a way to do this without adding anything to job_spec (set job._shell instead of /bin/bash directly in command= in ServiceBackend, much like LocalBackend), but you specified it so I added it, assuming you have a plan that needs it.